Verify add-on zip authenticity

yol opened the pull request:
Enable SSL connections for add-on zip downloads from the official repository and verify the downloaded files with the hashes provided in the redirect

(for the server side: cc @kib  @wsnipex)

Rationale
---
We've long been discussing how to make the Kodi update mechanism more secure (or rather, not *completely* insecure) with things like TUF etc.
TUF is unfortunately a very difficult path, so I would like to propose this as an interim "solution" for Leia.

It does not prevent all (or even most) of the bad things TUF would. But it at least doesn't make it so dead simple to own a Kodi installation via network and means we don't have to trust the mirrors as much as we do now.
The principle is: We get the hash and location of the zip file from the central mirror redirector run by ourselves (`mirrors.kodi.tv`). Since we do that over SSL, the hash cannot be tampered with. The redirector can then give us the URI of the file on a mirror of its choice, be it HTTP or HTTPS. Using SSL connections for all mirrors is not an option at the moment since many of those do not support it yet.

This is still WIP, but I want to get early feedback whether this is the direction we want to be going in first. The basic principle is implemented.

Problems
---
* If we decide to do this, it implies that we cannot later remove the mirror redirect and the `Content-Sha256` header without breaking updates in older installations.
* Third-party repositories do not directly benefit from this except when they duplicate our mirror setup. If they don't have too much traffic or can handle it, they should just use HTTPS. We might want to show a warning when a repository neither has this hash feature nor uses a SSL connection - or even make it unsupported?

TODO
---
- [x] Decide whether the old `.md5` code path should be kept (for 3rdparty repositories? does someone use it?)
- [x] Only enable the check when it is enabled for the repository
- [x] Fix base64 encoding in mirrorbits
- [x] Split `icondir` from `datadir` so Kodi doesn't use SSL connections for downloading the icons
- [x] Switch everything to sha256
- [x] Check whether we can secure `addons.xml` further and if it opens up any attack vectors
- [x] Wiki changes: https://kodi.wiki/view/Add-on_repositories

pkerling committed:
[curl] Enable SSL certificate validation

Not verifying any certificates renders SSL meaningless. cUrl should have
a stock of default CA certificates provided by the system/distributor.

pkerling committed:
[addons] Verify package authenticity via mirror redirector

pkerling committed:
[addons] Switch official repository datadir to https

pkerling committed:
Include guard cleanup

wsnipex commented:
much better then what we have now, so +1
is there a reason why you didn't go for sha256 directly?

pkerling commented:
> is there a reason why you didn't go for sha256 directly?

We don't (to the extent of my knowledge) have sha256 calculation code at the moment, so it's some work to integrate it first.

wsnipex commented:
right, openssl has this: https://www.openssl.org/docs/man1.0.2/crypto/EVP_DigestInit.html

pkerling commented:
Do we require OpenSSL at the moment and on all platforms?

wsnipex commented:
We still have an option to disable it, but we build and use it on all platforms.
If a standalone implementation of sha256 is a considerable overhead, we should consider making openssl required everywhere

Rechi commented on xbmc/addons/Repository.cpp:
```diff
@@ -55,6 +56,35 @@ using namespace KODI::MESSAGING;
+  // content-* headers are base64, convert to base16
+  std::string hash = StringUtils::ToHexadecimal(Base64::Decode(file.GetHttpHeader().GetValue("content-md5")));
+
+  if (location.empty() || hash.empty())
```
Not that I care too much about 3rd party repos, but this will break repos which do not redirect and repos not implementing `content-md5` in the response header.

pkerling replied:
This is on the TODO list ("Only enable the check when it is enabled for the repository").

You should see this PR more as a RFC on the general idea and not dive too deeply into the code yet :-)

pkerling commented:
Let's formally require OpenSSL if we - practically - do that already anyway. I might come back at you concerning the CMake changes once it becomes relevant.

tamland commented:
I think this is a really bad idea for the reasons you listed in 'problems'. Why not go for signatures distributed on mirrors? That has none of those disadvantages. Nor does it require https for security (just for privacy)

pkerling commented:
> Why not go for real signatures distributed on mirrors?

Signatures based on public-key schemes involve key management and require a solid system for key revocation - rolling out signatures without that is basically meaningless. It's complicated and error-prone (not to mention effort) to cook that up yourself. If we do that, it should be based on TUF or something comparable. TUF is difficult to integrate at the moment for known reasons. I can expand on that if you're interested.

That said, it would be nice to have solutions for the issues I listed. I think the 3rdparty repositories are not especially problematic. E.g. kodinerds (and most others distributing via GitHub I guess) use https anyway.
Regarding the commitment to keep the current mirror setup I'd like to invite @kib and @wsnipex if they have some ideas if we could alleviate that somehow. If we one day move to TUF or something else everything will change anyway though, so we have to keep that in mind.

wsnipex commented:
Infrastructure wise, it's not much of a commitment. This header can easily be added to normal web servers as well. That said, we should consider the current method of requesting the .md5 via extra request as fallback if the header is missing.This is trivial with any web server and also solves 3rd pt repos.

pkerling commented:
The infrastructure requirement is not only the header, but also that it is a redirect in the first place. If you decide to do the mirroring/load balancing differently without a redirect, it will not work.

So what about this:
* If we get a 3xx response:
  * If it has a Content-Sha256 header: use it
  * Otherwise: Try the `.md5` (or `.sha256` rather) method
* If we get a 200 response: Close the connection and return the original URL as download location. Means we have to connect twice but at least it'll continue to work without redirector.

pkerling commented:
Is it possible to set up mirrorbits so that it returns the redirect and headers also on `HEAD` request?

wsnipex commented on xbmc/addons/Repository.cpp:
```diff
@@ -55,6 +56,35 @@ using namespace KODI::MESSAGING;
+
+  // Return the location from the header so we don't have to look it up again
+  // (saves one request per addon install)
+  std::string location = file.GetHttpHeader().GetValue("location");
```
better use https://curl.haxx.se/libcurl/c/CURLINFO_REDIRECT_URL.html
and maybe also https://curl.haxx.se/libcurl/c/CURLOPT_HEADERFUNCTION.html

wsnipex replied:
I'd be possible to get all headers of all requests (incl 30x) while directly following the redirect.

pkerling replied:
We do not want to do that, at least with the current logic. We first get the hash, check if we have the file with that hash in the package cache already, and only if not download the file.

wsnipex commented:
The logic you outline above is what I had in mind, but by using the curlinfo_redirect_url function, we can avoid doing another connect. We still need a 2nd request for the .sha256, but that cannot be avoided and is OK as fallback method.

There is zero benefit in HEAD requests here, since mirrorbits still has to do the exact same work.

pkerling commented:
> by using the curlinfo_redirect_url function, we can avoid doing another connect

How so?

> There is zero benefit in HEAD requests here, since mirrorbits still has to do the exact same work.

I wasn't worrying about mirrorbits, I was worrying about the case when we ditch the mirror redirector. Then we would `GET` the URL to look up the hash  and instead of receiving a redirect as expected we get and buffer the data directly - but we cannot use it at that point and thus have to throw it away. That could be avoided by using `HEAD` always.

wsnipex commented:
>I wasn't worrying about mirrorbits, I was worrying about the case when we ditch the mirror redirector. Then we would GET the URL to look up the hash and instead of receiving a redirect as expected we get and buffer the data directly - but we cannot use it at that point and thus have to throw it away. That could be avoided by using HEAD always.

>Is it possible to set up mirrorbits so that it returns the redirect and headers also on HEAD request?

you asked about mirrorbits
For the normal webserver case, it should be possible to return the content heades on HEAD.
But let's not get too hung up on this, since it's not likely we ditch mirrorbits anytime soon, since it's working great

tamland commented:
>Signatures based on public-key schemes involve key management and require a solid system for key revocation - rolling out signatures without that is basically meaningless. It's complicated and error-prone (not to mention effort) to cook that up yourself. If we do that, it should be based on TUF or something comparable. TUF is difficult to integrate at the moment for known reasons. I can expand on that if you're interested.

I can think of many reason because from a quick look it's a huge and complete framework made to cover all sorts of use cases, many of which we don't have. That doesn't mean we cannot implement part of it though. If you're worried about 'cooking stuff up yourself' this custom redirect scheme is exactly that. At least signing is a widely understood and solve problem.

MartijnKaijser commented:
@tamland who is going to implement that signings stuff and when? Yes that's all wishful thinking however we can debate everything to death and go nowhere or take this "easy" improved solution until some one finds the time to take the extra effort.

tamland commented:
> @tamland  who is going to implement that signings stuff and when? Yes that's all  wishful thinking however we can debate everything to death and go  nowhere or take this "easy" improved solution until some one finds the  time to take the extra effort.

Maybe I will, but if people are against the very idea I won't waste my time on it.

MartijnKaijser commented:
Nowhere was it mentioned you would be willing to do it and then again is the question when?

tamland commented:
This isn't an either x or y thing. It only becomes that when we require hashes and redirects, or require that security isn't removed again (which was discussed above). If it doesn't lead to hacking it's never an issue to take an improved solution until there's an alternative.

pkerling commented:
> I can think of many reason because from a quick look it's a huge and complete framework made to cover all sorts of use cases, many of which we don't have.

I was talking more about technical reasons. TUF  is not that huge actually and it covers pretty much *exactly* our use case :-)

> If you're worried about 'cooking stuff up yourself' this custom redirect scheme is exactly that. 

In principle yes, but the complexity and screw-up potential is hardly comparable to signatures and key management.

> This isn't an either x or y thing. It only becomes that when we require hashes and redirects, or require that security isn't removed again (which was discussed above)

We already do exactly that (require hashes and that they not be removed in the future) with the `.md5` files. This is a simple extension of that scheme that will include a fallback to the status quo as discussed - with the proposed modifications the redirect is not required any more. So effectively, the only thing that changes and cannot easily be changed back is that we **require SSL connections for the zip requests** since that will be hardcoded in the `addon.xml`.

> Maybe I will, but if people are against the very idea I won't waste my time on it.

My personal opinion is that the time would be better spent integrating TUF (or something similar, I'm not that fixated on TUF actually).

pkerling committed:
[addons] Refactor CRepository::FromExtension

pkerling committed:
[addons] Allow specifying separate URL for add-on art

Needed so we do not overload the mirror redirector with SSL requests
(when datadir is set to an HTTPS URL)

tamland commented:
> We already do exactly that (require hashes and that they not be removed in the future) with the .md5 files. This is a simple extension of that scheme that will include a fallback to the status quo as discussed - with the proposed modifications the redirect is not required any more. So effectively, the only thing that changes and cannot easily be changed back is that we require SSL connections for the zip requests since that will be hardcoded in the addon.xml.

Then I don't have any issues with it. Technically you could probably issue a self update of the repository add-on to remove tls but imo we *should* commit to tls for privacy regardless of which solution is chosen.

> My personal opinion is that the time would be better spent integrating TUF (or something similar, I'm not that fixated on TUF actually).

But you said 'for known reasons' it's difficult to integrate..

tamland commented:
> It does not prevent all (or even most) of the bad things TUF would. But it at least doesn't make it so dead simple to own a Kodi installation via network and means we don't have to trust the mirrors as much as we do now.

Could you elaborate what you mean this method doesn't protect again? Because looking at the 'goals' list I'm really unsure what insecure about it (assuming that the mirror list and addons.xml is secured and protected by us).

pkerling commented:
> But you said 'for known reasons' it's difficult to integrate..

There were some discussions, but they are scattered around Slack/forum/in-person convos, so I might have made too many assumptions there. I think what I wrote on the GSoC wiki page sums up the most prominent problem: "TUF does not currently have a C or C++ implementation. The reference implementation is in Python, which we do not want to integrate with C++ for this core piece of application code. Due to security concerns, we also do not want to implement TUF ourselves. That leaves using rust-tuf, an experimental Rust implementation."
So we would have to write, maintain, and use a C wrapper for rust-tuf, which is still experimental.

> Could you elaborate what you mean this method doesn't protect again? Because looking at the 'goals' list I'm really unsure what's insecure about it

See section 1.5.2 of https://github.com/theupdateframework/specification/blob/master/tuf-spec.md
Also note that TUF does not rely on SSL for security at all.

> (assuming that the mirror list and addons.xml is secured and protected by us).

`addons.xml` is not secured by this proposal. The reason is that `addons.xml` sees by far the most requests and I've been told by @kib that they are not sure whether the server could handle that amount of SSL connections.

But I do think that the change I'm proposing gets us a long way from the current state of affairs in terms of update security with very little effort - that's why I'm doing it :-)

tamland commented:
> See section 1.5.2 of https://github.com/theupdateframework/specification/blob/master/tuf-spec.md
Also note that TUF does not rely on SSL for security at all.

I mean if we already decide to go for tls anyway for other season; do we really get anything from this additional layer?

>addons.xml is not secured by this proposal. The reason is that addons.xml sees by far the most requests and I've been told by kib that they are not sure whether the server could handle that amount of SSL connections.
>
>But I do think that the change I'm proposing gets us a long way from the current state of affairs in terms of update security with very little effort - that's why I'm doing it :-)

That's why I've been preaching over and over again to get away from a routing everything through a central entrypoint;) Every time it's blocking the features we want, now including adding proper security, because it's too much traffic to handle. Sorry if this came out wrong previously. I wasn't trying to be negative about any security improvements. It's just a fear of digging the hole deeper and making it even harder to get away and fix that 'real' underlying problem.

pkerling commented:
> I mean if we already decide to go for tls anyway for other season; do we really get anything from this additional layer?

Yes. As said previously, TUF fixes problems TLS does not. They might not seem as big as the "any random guy on the network can push arbitrary code to Kodi" problem we have at the moment, but still worthwhile to solve in my opinion. It would also allow a move to mirror decentralisation (see below).

> That's why I've been preaching over and over again to get away from a routing everything through a central entrypoint;)

I'm all for that actually., but it means that the mirrors are accessed directly, via DNS anycast or mirror list. That's only possible if we do TUF or a custom PKI like you propose. Otherwise we would have to completely trust the mirrors (TLS won't help obviously), which is not good.

pkerling committed:
[addons] Switch file hashes to sha256

MD5 is known to be insecure for some time now.

pkerling commented:
@chewitt I've seen that the LibreELEC repositories use plain HTTP at the moment. Do you think it would be possible that you enable SSL at least for the zip files? This PR actually doesn't make that much sense without at least getting you guys on board.

pkerling committed:
Add CCurlFile::GetRedirectURL()

pkerling committed:
[addons] Validate addon identifier

pkerling committed:
[addons] Validate addons.xml against provided hash

tamland commented:
> I'm all for that actually., but it means that the mirrors are accessed directly, via DNS anycast or mirror list. That's only possible if we do TUF or a custom PKI like you propose. Otherwise we would have to completely trust the mirrors (TLS won't help obviously), which is not good.

Don't know where you got that from. Definitely never propose that. You ofc use existing pki for keys

pkerling commented:
> Don't know where you got that from. Definitely never proposed that. You ofc use existing pki for keys

What do you mean with "existing PKI?"

pkerling committed:
[addons] Validate addon version

pkerling commented:
Hash change to sha256 is now done. Two questions:

* Do we want to keep .md5 for compatibility?
* I've modified the code to actually check the hash for `addons.xml` (so our mirrors can't replace the file). We never required the hash to actually match previously, it's even written on the wiki page. Is it OK to change that again?

tamland commented:
> What do you mean with "existing PKI?"

As in you simply host the keys on a trusted site. Then there is no key management and no need to run your own custom pki like TUF requires you to.

tamland commented:
>- Do we want to keep .md5 for compatibility?
>- I've modified the code to actually check the hash for addons.xml (so our mirrors can't replace the file). We never required the hash to actually match previously, it's even written on the wiki page. Is it OK to change that again?

Imo md5 *verification* can just be dropped because it hardly servers any purpose anymore, but *not* the 'checksum' file. This is and has always been used for change detection only, not verification. These are different things and changing that breaks all existing repos. And already sure repos have no need for the latter.

tamland commented on xbmc/addons/Repository.cpp:
```diff
@@ -52,9 +55,83 @@
+    if (!FetchChecksum(path + ".sha256", hash))
+    {
+      CLog::Log(LOGERROR, "Failed to find hash for {} from HTTP header and in separate file", path);
+      return {};
```
You can't just change this to fail if the .sha256 isn't there. That breaks all existing repos that has set 'hashes' to true.

tamland replied:
Also, hard-coding 'sha256' is not a good idea.. It might break tomorrow and you have the md5 situation again. I suggest putting the alg. in `<hashes>` for instance. That might solve the backwards compatibility issue as well, as you can just let 'true' mean md5.

pkerling replied:
> That breaks all existing repos that has set 'hashes' to true.

How common is that? (out of curiosity)

> I suggest putting the alg. in <hashes> for instance. That might solve the backwards compatibility issue as well, as you can just let 'true' mean md5.

So the other repositories using MD5 hashes, what purpose are they using them for? Is there even any meaning in checking those hashes at all?

pkerling replied:
Or is that what you meant with "MD5 verification can be dropped?" :-)

pkerling commented:
> Imo md5 verification can just be dropped because it hardly servers any purpose anymore

The purpose would be that the hash comes from our server so mirrors cannot replace `addons.xml` with modified versions (which they can do now).

tamland commented:
>> Imo md5 verification can just be dropped because it hardly servers any purpose anymore
>>
> The purpose would be that the hash comes from our server so mirrors cannot replace addons.xml with modified versions (which they can do now).

The keyword here is md5. md5 verification 'hardly servers any purpose'

pkerling commented:
> The keyword here is md5. md5 verification 'hardly servers any purpose'

Makes sense, got a bit confused which part specifically you were talking about.

> These are different things and changing that breaks all existing repos. And already sure repos have no need for the latter at all.

Yes, I think the `addons.xml` verification mostly affects the special case of our own repo. So should we make it opt-in?

pkerling commented:
@tamland This should address your concerns.

wsnipex commented on addons/repository.xbmc.org/addon.xml:
```diff
@@ -8,10 +8,10 @@
 	<extension point="xbmc.addon.repository">
 		<info>http://mirrors.kodi.tv/addons/leia/addons.xml.gz</info>
-		<checksum>http://mirrors.kodi.tv/addons/leia/addons.xml.gz.md5</checksum>
+		<checksum>http://mirrors.kodi.tv/addons/leia/addons.xml.gz.sha256</checksum>
```
can you change this to http://mirrors.kodi.tv/addons/leia/addons.xml.gz?sha256
which is the actual url mirrorbits uses. Saves 1 rewrite on the webserver.

pkerling replied:
> Can we use the url from <info> and append ?sha256

We can, but that would break 3rd-party repos. Do you have a strong reason to change that?

pkerling commented:
jenkins build this please

Rechi commented on xbmc/addons/AddonInstaller.cpp:
```diff
@@ -524,12 +516,28 @@ bool CAddonInstallJob::DoWork()
+      std::string path{m_addon->Path()};
+      std::string hash;
+      KODI::UTILITY::CDigest::Type hashType;
+      if (m_repo) {
```
curly bracket has to go into a new line

Rechi commented on xbmc/addons/AddonInstaller.cpp:
```diff
@@ -565,20 +575,20 @@ bool CAddonInstallJob::DoWork()
       {
-        md5 = CUtil::GetFileDigest(package, KODI::UTILITY::CDigest::Type::MD5);
-        if (!StringUtils::EqualsNoCase(md5, m_hash))
+        std::string actualHash = CUtil::GetFileDigest(package, hashType);
```
`hashType` may be used uninitialized

pkerling replied:
either both `hash` and `hashType` are set or neither is

pkerling replied:
OK so the compiler doesn't understand this... I'll see what I can do.

pkerling commented:
jenkins build this please

pkerling committed:
Add TypedDigest utility

pkerling committed:
[addons] Utilize TypedDigest

pkerling commented:
Oh seems I accidentally used a fmt-4 feature. Will have to remove that.

pkerling commented:
So... for me this is good to go. Will merge on Monday if there are no objections.

pkerling commented:
Running in Milhouse build for a few days now without obvious issues.

@chewitt Ping me when you've done the adjustments for LibreELEC
